chainer/functions/array/as_strided.py

```diff
@@ -0,0 +1,387 @@
+
+        if out_maybe_overlap:
+            out_indices = _stride_array(flatten_full_indices, out_shape,
+                                        out_strides,
+                                        out_storage_offset - common_min_ptr)
+            _index_add(storage, out_indices, gy)
+        else:
+            storage_view = _stride_array(storage, out_shape, out_strides,
+                                         out_storage_offset - common_min_ptr)
+            storage_view[:] = gy[:]
+
+        inp_maybe_overlap = _maybe_overlapping_memory(inp_shape, inp_strides)
+        if inp_maybe_overlap:
+            count = xp.zeros_like(storage)
+            inp_indices = _stride_array(flatten_full_indices, inp_shape,
+                                        inp_strides,
+                                        inp_storage_offset - common_min_ptr)
+            _index_add(count, inp_indices, xp.ones(1))
+            with np.errstate(divide='ignore', invalid='ignore'):
+                storage /= count
+
+        return _stride_array(storage, inp_shape, inp_strides,
+                             inp_storage_offset - common_min_ptr),
+
+    def backward(self, target_input_indexes, grad_outputs):
+        raise NotImplementedError
+
+
+def as_strided(x, shape, strides, storage_offset=None):
+    """Create a new view of array with the given shape, strides, and offset.
+
+    Args:
+        x (tuple of :class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`):
+            The array pointing a memory buffer. Its view is totally ignored.
+        shape (tuple of int):
+            The shape of output.
+        strides (tuple of int):
+            The strides of output, given in the unit of steps.
+        storage_offset (int):
+            The offset between the head of allocated memory and the pointer of
+            first element, given in the unit of steps.
+
+    Returns:
+        ~chainer.Variable: The strided variable.
+
+    .. warning::
+        Users should be aware that this function potentially causes unintended
+        side effects. See `numpy.lib.stride_tricks.as_strided`_ for the detail.
+
+    .. note::
+        The backward algorithm is borrowed from `torch.Tensor.as_strided`.
+        Therefore, the returned gradient of ``backward`` is *layout-agnostic*
+        when ``x`` contains memory overlap. See notes in pytorch's source
+        code (as_strided Backward and layout-aware/agnostic autograd) too.
+
+    .. note::
+        In this function ``strides`` and ``storage_offset`` are given in the
+        unit of steps instead of bytes. This specification differs from
+        :func:`numpy.lib.stride_tricks.as_strided`.
+
+    .. admonition:: Example
+
+        >>> from chainer import functions as F, Variable
+        >>> x = Variable(np.arange(4, dtype=np.float32))
+        >>> x
+        variable([0., 1., 2., 3.])
+        >>> y = F.as_strided(x, (3, 2), (4, 4), 0)
```

Comment:
I got error result.
```
Failed example:
    y = F.as_strided(x, (3, 2), (4, 4), 0)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest default[3]>", line 1, in <module>
        y = F.as_strided(x, (3, 2), (4, 4), 0)
      File "/work/chainer/chainer/functions/array/as_strided.py", line 387, in as_strided
        return AsStrided(shape, strides, storage_offset).apply((x,))[0]
      File "/work/chainer/chainer/function_node.py", line 318, in apply
        outputs = self.forward(in_data)
      File "/work/chainer/chainer/functions/array/as_strided.py", line 222, in forward
        return _stride_array(x, self.shape, self.strides, self.storage_offset),
      File "/work/chainer/chainer/functions/array/as_strided.py", line 168, in _stride_array
        raise ValueError("Out of buffer: too large index was specified")
    ValueError: Out of buffer: too large index was specified```

Reply:
fixed.
I'm sorry that this example was for the prototype of `F.as_strided`. After I wrote the example, I changed the specification, but I forgot to change the example code.